### DISCUSSION TITLE
Form data in custom validators

### DISCUSSION THREAD

**Discussion by Aftabnack** [react-hook-form/react-hook-form#4517] [Ideas] [answered]
I love this library and we have paired it well with Material UI to build our application.
I have one feature request which I feel will benefit a lot of usecases.

#### Current Situation

* Currently for doing custom validation I can pass a function which has to return either an error message or return true if it is valid.
* If my validation depends on value of some other field, I am currently having to use `watch` method or `useWatch` hook and use that values like so.
```js
function CustomComponent(props) {
  const { control } = props;

  const field2Value = useWatch({ control, name: "field2" });

  const {
    field: { ref, ...rest },
  } = useController({
    name: "field1",
    control,
    rules: {
      validate: (value) => {
        if (value < field2Value) {
          return "This value cannot be lesser than field2";
        }
        return true;
      },
    },
  });

  return <Input ref={ref} {...rest} />;
}
```
* But notice here, I need to create a separate component so that the form doesn't re-render. Since useWatch will re-render the component

#### Suggestion

* Can we add the current serialized object from the form to the validation signature??
* This will help in preventing to use useWatch, separating it into a different component to prevent re-renders.
* Overall the userland code will be very concise!!

```js
function CustomComponent(props) {
  const { control } = props;

  const {
    field: { ref, ...rest },
  } = useController({
    name: "field1",
    control,
    rules: {
      validate: (value, formData) => {
        if (value < formData.field2) {
          return "This value cannot be lesser than field2";
        }
        return true;
      },
    },
  });

  return <Input ref={ref} {...rest} />;
}
```

**Comment by bluebill1049** [comment, marked as answer]
you can use `getValues()` inside instead of watch, we did consider performance reason not to include as second arg.

**Reply by Aftabnack** [reply]
I will give it a shot with getValues, it’s great if that works without causing rerenders.
I think it should, I’ll keep you posted.

Thanks!!

**Reply by Aftabnack** [reply]
Thanks @bluebill1049, it worked.
I didn't have a requirement of watching values, `getValues()` did the trick for me.

**Reply by bluebill1049** [reply]
no worries.
